[CUB][device] Add a env-based overload of the device segmented reductions primitives

rbourgeois33 opened the pull request:
## Description

closes https://github.com/NVIDIA/cccl/issues/6673

This Draft PR  implements a env-based overload of the device segmented reductions primitives.

@gevtushenko, following [our discussion](https://github.com/NVIDIA/cccl/discussions/6497#discussioncomment-14961319):

> Consider adding a static assert that if determinism is specified - it's run-to-run or not-guaranteed.

I have checked that the required determinism, if provided, must not be `gpu_to_gpu`. As far as I understand, it is equivalent to check that it's run-to-run or not-guaranteed.

I made this check list of what this PR should verify before being merged below. Let me know what I should add to the `Sum` example to tick the first box :).

## Checklist
- [x] We have agreed on the way to implement the env-based device segmented reductions starting with the first one:`Sum`.
- [ ] I have extended the overload to other relevant primitives.
- [ ] The documentation is up to date with these changes.
- [ ] If requirements are the same as [the CUDALibrarySamples repo](https://github.com/NVIDIA/CUDALibrarySamples), I have updated the licence headers and signed my commits.
- [ ] I have squashed commits

srinivasyadav18 reviewed (changes requested):
Thank you for trying to contribute to CCCL! Apologies for very early review on the draft. I have some suggestions which might help you to get better context. Feel free to ask any questions you have.

rbourgeois33 commented:
Hello @srinivasyadav18 . Thanks a lot for the feedback :)

I added several tests in the new file `cub/test/catch2_test_device_segmented_reduce_env_api.cu` as you suggested.

These include three tests to verify that the new API correctly accepts a stream and the two non-`gpu_to_gpu` determinism arguments from the `env`.

In the comment header for the new function, I linked to the `segmented-reduce-sum-env-stream example`, since the determinism arguments are currently unused. I believe this is more useful for the documentation. Perhaps the fact that the determinism arguments are unused should be clearly stated.

I removed my example code because could not  stay here long-term. I also didn’t add a tests for passing an `mr` argument since memory resources are still experimental.

Regarding the complete boiler plate to extract useful info from `env`, I think it's a great idea ! Less code duplicate is always better, in particular since the `env` extraction is quite verbose. I subscribed the PR and make changes accordingly in time.

Cheers !

On a side note, looking at 

https://github.com/NVIDIA/cccl/blob/38162faf37deea00c6e70366cb789956dae3fb28/cub/test/catch2_test_device_reduce_env_api.cu#L87

it seems to me that it is an incomplete duplicate of:

https://github.com/NVIDIA/cccl/blob/38162faf37deea00c6e70366cb789956dae3fb28/cub/test/catch2_test_device_reduce_env_api.cu#L148 since it has the same title and does not check correctness of the result. Perhaps it should be removed.

srinivasyadav18 commented on cub/cub/device/device_segmented_reduce.cuh:
```diff
@@ -407,6 +407,184 @@ struct DeviceSegmentedReduce
+      ::cuda::std::execution::__query_result_or_t<requirements_t, //
+                                                  ::cuda::execution::determinism::__get_determinism_t,
+                                                  ::cuda::execution::determinism::run_to_run_t>;
+
```
currently, tunings are not extracted from the `env`. It is optional, but if would like to, you could extract it from `env` and make a `policy_hub` out of it, and pass it on to the `Dispatch`, it would be really great.

srinivasyadav18 replied:
Looks great! Please proceed with adding the unit test.

srinivasyadav18 reviewed (commented):
These two comments are totally optional, feel free to look into. 
Please let me know if you have questions. Thanks!

rbourgeois33 commented on cub/test/catch2_test_device_segmented_reduce_env_api.cu:
```diff
@@ -0,0 +1,77 @@
+  c2h::device_vector<int> d_out(3);
+
+  auto env = cuda::execution::require(cuda::execution::determinism::not_guaranteed);
+
```
Let's make a separate example out of this and refer to it in the function's header as it shows a richer api.

rbourgeois33 commented:
Just added another unit test. Let me know if you think it would be useful or not to extend the work to other segmented reductions.

srinivasyadav18 commented on cub/test/catch2_test_device_segmented_reduce_env.cu:
```diff
@@ -0,0 +1,77 @@
+
+C2H_TEST("Device segmented sum can be tuned", "[reduce][device]", block_sizes)
+{
+  constexpr int target_block_size = c2h::get<0, TestType>::value;
```
With the current test logic, we are not sure if `target_block_size` is being used or not. Can you please in-corporate `target_block_size` logic into num_segments and num_items. for e.g, have one segment (with `target_block_size` as segment size) with all elements initialized to 1, and then compare the result i.e segment sum with block size.

srinivasyadav18 replied:
@rbourgeois33 Adding the above suggestion does not actually help in testing if `target_block_size` is used in the implementation. For e.g. [device_reduce_env.cu](https://github.com/NVIDIA/cccl/blob/main/cub/test/catch2_test_device_reduce_env.cu#L140) uses a special custom functor to test this logic. We cannot do the same here as this is `Sum()` algorithm and we cannot pass any reduction operators. Apologies for the confusion. We can skip adding this logic.

NaderAlAwar reviewed (approved):
Great work @rbourgeois33! We still need to create the env based overloads for the other primitives.
